[14기 송지원] step2 - 상태 관리로 메뉴 관리하기

zwonlala opened the pull request:
## 🎯 step2 요구사항 - 상태 관리로 메뉴 관리하기

- [X] [localStorage](https://developer.mozilla.org/ko/docs/Web/API/Window/localStorage)에 데이터를 저장하여 새로고침해도 데이터가 남아있게 한다.
- [X] 에스프레소, 프라푸치노, 블렌디드, 티바나, 디저트 각각의 종류별로 메뉴판을 관리할 수 있게 만든다.
  - [X] 페이지에 최초로 접근할 때는 에스프레소 메뉴가 먼저 보이게 한다.
- [X] 품절 상태인 경우를 보여줄 수 있게, 품절 버튼을 추가하고 `sold-out` class를 추가하여 상태를 변경한다.
- 품절 상태 메뉴의 마크업

```js
<li class="menu-list-item d-flex items-center py-2">
  <span class="w-100 pl-2 menu-name sold-out">${name}</span>
  <button
    type="button"
    class="bg-gray-50 text-gray-500 text-sm mr-1 menu-sold-out-button"
  >
    품절
  </button>
  <button
    type="button"
    class="bg-gray-50 text-gray-500 text-sm mr-1 menu-edit-button"
  >
    수정
  </button>
  <button
    type="button"
    class="bg-gray-50 text-gray-500 text-sm menu-remove-button"
  >
    삭제
  </button>
</li>
```

<br>
<br>

## 😸 step2 코드리뷰 확인 사항  

1. 코드 불필요한 주석은 없는가? 
2. 불필요한 “console.log()” 는 없는가? 
3. eol 개행을 하였는가? 
	- [파일 끝에 개행을 추가해야 하는 이유](https://coderifleman.tumblr.com/post/115464362564/%ED%8C%8C%EC%9D%BC-%EB%81%9D%EC%97%90-%EA%B0%9C%ED%96%89%EC%9D%84-%EC%B6%94%EA%B0%80%ED%95%B4%EC%95%BC-%ED%95%98%EB%8A%94-%EC%9D%B4%EC%9C%A0)
	- [파일마다 EOL(End Of Line)을 왜 넣어야 할까](https://avocado12-note.tistory.com/11)
4. var 사용을 지양하고 let, const를 적절히 사용하였는가? 
5. 불필요한 전역변수를 사용하지 않았는가? 
6. 변수명에 적절한 네이밍을 적용하였는가?
7. 중복코드를 최소화하였는가?
8. 하나의 함수가 너무 많은 일을 하고 있진 않은가?
9. UI 렌더링 코드와 비즈니스 로직 코드를 분리하였는가?
10. [부정 조건을 사용하지 않는가?](https://velog.io/@seob/JavaScript-Clean-Code-%EA%B0%80%EB%8F%85%EC%84%B1-%EB%86%92%EC%9D%B4%EA%B8%B0)
11. [Early return pattern을 적용하였는가?](https://uminoh.tistory.com/m/35)

<br>
<br>

## 👨‍💻 코드 설명

### [LocalStorageUtil.js](https://github.com/blackcoffee-study/moonbucks-menu/blob/7d01dd65b339b63e2c2eccc22462725fefb9d093/src/util/LocalStorageUtil.js)

- localStorage에 저장 / 불러오는 로직(함수) 정의된 class

<br>

### [StringUtil.js](https://github.com/blackcoffee-study/moonbucks-menu/blob/7d01dd65b339b63e2c2eccc22462725fefb9d093/src/util/StringUtil.js)

- 메뉴의 id를 생성하기 위해 문자열을 입력받아, 해당 값을 기반으로 hash 값을 생성하여 리턴하는 함수가 정의된 클래스

<br>

### [MenuType.js](https://github.com/blackcoffee-study/moonbucks-menu/blob/7d01dd65b339b63e2c2eccc22462725fefb9d093/src/model/MenuType.js)

- 메뉴타입에 대한 상수 정의
- 메뉴별 한글 문자열 상수 정의
- 모든 메뉴 타입을 리턴하는 함수 
- 메뉴 타입에 해당하는 한글 문자열 리턴하는 함수
- 메뉴타입인지 판별하는 함수

정의된 파일

<br>
<br>

## 🤔 고민되는 부분

- 


<br>
<br>

리뷰해주시면 감사하겠습니다!! 🙏

zwonlala committed:
feat: Menu listItem에 품절 상태 관련 구현

메뉴 별 <li>에
- 품절 시 취소선 표시되는 UI
- 품절 버튼 추가

zwonlala committed:
feat: "품절" 버튼 클릭시 상태 및 UI 업데이트 되는 로직 추가

zwonlala committed:
refactor: '메뉴 이름' 업데이트시에도 updateMenuListElementView() 함수 통해 UI 업데이트 되게 수정

zwonlala committed:
refactor: "품절" 관련 네이밍 "purchasble" -> "soldout" 으로 변경

zwonlala committed:
feat: 선택된 메뉴 타입 관련 된 파일 MenuType.js 정의

zwonlala committed:
feat: 추가된 MenuType을 Menu class 와 MenuStatus class, index.js에 적용

zwonlala committed:
feat: 메뉴판 상단 메뉴 타입 버튼 클릭시 동작되는 로직 구현

zwonlala committed:
feat: LocalStorage 관련 저장, 로드 로직 모여있는 LocalStorageUtil.js 정의

zwonlala committed:
feat: LocalStorageUtil 사용하여 데이터 저장 및 불러오는 로직 구현

dhrod0325 commented on src/js/index.js:
```diff
@@ -0,0 +1,191 @@
+
+function onSubmitForm(e) {
+	e.preventDefault();
+	const inputString = String(inputTag.value).trim();
```
`nputTag.value` 는 원래 `String`인데 `trim`을 사용하시기 위해 `String` 으로 선언 하신 걸까요~? 
`optional` 을 이용해서 좀 더 심플하게 작성 할 수 있을 것 같아요.

```suggestion
	const inputString = inputTag.value?.trim();
```

 value가 undefined 이거나 null 인 경우에는 trim 메소드가 실행되지 않으니 적지만 성능도 좀 더 좋아지겠네요!

https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Operators/Optional_chaining

zwonlala replied:
오 넵넵

> nputTag.value 는 원래 �String인데 trim을 사용하시기 위해 String 으로 선언 하신 걸까요~?
> ㄴ 넵넵 맞습니다! undefind 값 올까봐 String()으로 한번 묶어줬습니다!

그렇네요 optional 적용하면 더 깔끔하게 코드 작성 가능하네요!
감사합니다! :)

dhrod0325 commented on src/js/index.js:
```diff
@@ -0,0 +1,191 @@
+	}
+
+	// 현재 state의 해당하는 Menu 배열을 불러와, 각각의 menu를 기반으로 <li> Element를 생성하여 menuList에 삽입
+	state.getMenuList().forEach(
+		(menu, index) => menuList.insertAdjacentHTML("beforeend", getMenuListItemHTMLString(menu, index))
```
```suggestion
function renderMenuItem(menu,index){
    menuList.insertAdjacentHTML("beforeend", getMenuListItemHTMLString(menu, index));
}

state.getMenuList().forEach(renderMenuItem);
```

메뉴 아이템을 추가하는 부분을 분리하면 좀 더 코드 읽기가 쉬울 것 같아요~!

zwonlala replied:
아하 메뉴 리스트 그려주는 부분을

- (사전 작업) + 메뉴 리스트 그려주는 부분
- 메뉴 리스트 아이템 하나 그려주는 부분

이렇게 분리하는 방향을 말씀해주신 거군요!

"함수는 하나의 일만 한다는 원칙"에 잘 부합하는 것 같네요!

감사합니다! :)

dhrod0325 commented on src/util/LocalStorageUtil.js:
```diff
@@ -0,0 +1,11 @@
+class LocalStorageUtil {
+    static KEY = 'MOONBUCKS_MENU';
+
+    static store(value) {
+        window.localStorage.setItem(LocalStorageUtil.KEY, value);
+    }
+
+    static load() {
+        return window.localStorage.getItem(LocalStorageUtil.KEY);
+    }
+}
```
`static`으로만 이루어진 유틸 클래스를 함수가 아니고 클래스로 작성하신 이유가 궁금해요~! 😎

zwonlala replied:
@dhrod0325 음... 개인적으로 `바닐라 TS` && `class 문법`을 사용하여 구현하는 방향이 익숙해서 이런식으로 작업을 진행하였습니다!

몇가지 공통적인 목적을 가진 함수들을 하나로 묶어서 관리해주고 싶었는데,

개인적으로 object 내에 함수를 정의해서 관리하는 구조 보다는 
class 문법을 사용하여 관리해주는 구조가 더 좋아보여서요~

~~(물론 그냥 class 문법은 syntactical sugar 지만요..!)~~
> 오 댓글 달면서 구글링 해봤는데 [ES6 Class는 단지 prototype 상속의 문법설탕일 뿐인가?](https://roy-jung.github.io/161007_is-class-only-a-syntactic-sugar/) 이런 글이 있네요 🤔

zwonlala replied:
@dhrod0325 혹시 다른 더 좋은 구조나 방법? 이 생각나신 걸까요? 😄

dhrod0325 replied:
저도 백엔드 개발자라 처음엔 모두 class문법으로 관리하고 싶어서 `static` 을 자주 사용했었는데 `static util class`를 사용할거면 그냥 함수를 쓰는게 더 나은거 같은 느낌이 들더라구요..! 

그냥 궁금해서 여쭤본거에요! ㅎㅎ

zwonlala replied:
아하 그렇군요~
좋은 의견 감사합니다! :) 

>  static util class를 사용할거면 그냥 함수를 쓰는게 더 나은거 같은 느낌이 들더라구요..!

이렇게 생각하게 되신 사례? 계기? 그런게 있나요?

dhrod0325 commented on src/model/MenuStatus.js:
```diff
@@ -0,0 +1,115 @@
+class MenuStatus {
```
비지니스 로직이 따로 분리되어 있어서 좋네요~! 그런데 `console` 로그가 많아서 코드를 보고 로직을 읽어나가는데 약간 방해되는 느낌인 것 같아요.! 로그 찍는 부분도 메서드로 바꿔서 분리하면 어떨까요~?

zwonlala replied:
@dhrod0325 오 넵넵
저도 구현하면서 핵심 부분보다 예외처리한 부분의 양이 너무 많아
해당코드의 핵심을 보는데 가독성을 해친다는 생각을 했었는데 🤔 
좋은 방향인 것 같네요!

흠.. 그런데 어떤 방식으로 메소드로 분리할지 잘 감이 안잡히는데 
혹시 힌트나 설아아빠님? 께서는 어떤 방향으로 진행하실지 알려주실 수 있나요? 😃

dhrod0325 replied:
https://github.com/blackcoffee-study/moonbucks-menu/pull/290/files#r933356456

저라면 이렇게 분리 할 것 같아요!

dhrod0325 reviewed (commented):
안녕하세요 지원님!

리뷰 몇개 달아 보았어요! 😀

zwonlala committed:
feat: hash 값 생성하는 로직, Menu 클래스 생성자에 id 값 설정해주는 로직 추가

zwonlala committed:
fix: index 값 기반으로 메뉴 수정/삭제하여 발생하던 버그 수정

기존 index 값 기준으로 메뉴 업데이트 && 삭제되던 로직 id 값 기반으로 변경되게 수정

zwonlala committed:
refactor: <input>의 value trim()하는 로직에 옵셔널 체이닝 적용된 구조로 변경

zwonlala committed:
refactor: 각 메뉴 <li> 그려주는 부분 함수화

메뉴를 그려주는 부분을

- 기존 메뉴 리스트 삭제 + 각 메뉴 별로 render 함수 호출
- 메뉴 리스트 아이템 하나 render 하는 로직

으로 분리하여 구현

zwonlala committed:
refactor: 각 메뉴 이름에 아이콘 추가

zwonlala committed:
refactor: js 파일간 import / export 하는 구조로 변경

dhrod0325 commented on src/model/MenuStatus.js:
```diff
@@ -17,19 +28,54 @@ class MenuStatus {
-    update(index, newMenu) {
-        if (typeof(index) !== 'number') {
-            console.error(`Invalid Input! ${index} is not 'number' value! @MenuStatus.update() 1st parameter`);
+    setSelectedMenuType(newMenuType) {
+        if (!newMenuType) {
+			console.error(`Invalid Input! ${newMenuType} is 'undefined' or 'null' or ""(empty string) value! @MenuStatus.setSelectedMenuType() 1st parameter`);
+            return;
+		}
+		if (!MenuTypeUtil.isMenuType(newMenuType)) {
+			console.error(`Invalid Input! ${newMenuType} is not value declared at "MenuType.js" file! @MenuStatus.setSelectedMenuType() 1st parameter`);
+            return;
+		}
+
+        this.selectedMenuType = newMenuType;
+        this.store();
+    }
```
```suggestion
    setSelectedMenuType(newMenuType) {
        if (!this.validateMenuType(newMenuType)) {
	   return;
        }

        this.selectedMenuType = newMenuType;
        this.store();
    }
```

이렇게 validate함수로 빼면 아래 로직을 읽어나가기가 쉬워지지 않을까요~?

zwonlala replied:
와 너무 좋은 것 같네요!!! 🤩 

좋은 인사이트를 주신 것 감사합니다! 🙇‍♂️ 

그런데 `validateMenuType()` 라는 함수를 MenuStatus 클래스에서 this로 접근하기 보단

`MenuType` 클래스 내에 static 함수로 정의하여 사용하는게 더 좋아보여서

제가 정의한 각 클래스 내부에 입력한 값이 유효한지 검증하는 static 함수를 추가로 정의하려 합니다!


좋은 포인트 코멘트 해주셔서 감사해요! :)

zwonlala committed:
refactor: 각 모델에 정의된 함수의 파라미터 유효한 값인지 판단하는 로직 함수화하여 중복 코드 제거

bingwer commented on src/model/MenuType.js:
```diff
@@ -0,0 +1,44 @@
+});
+
+
+export {MenuType, MenuTypeUtil};
```
이 파일과 `LocalStorageUitll.js`, `StringUitl.js`파일에 eol 개행이 빠져있는거 같아요~

zwonlala replied:
아 넵넵 놓친 부분 찾아주셔서 감사합니다! :)

bingwer commented on src/js/index.js:
```diff
@@ -1,18 +1,25 @@
 const menuCount = document.querySelector(".menu-count");
-
+const menuHeader = document.querySelector(".heading h2");
+const menuTypeSelectButtonWarpper = document.querySelector("header nav");
```
중요한건 아니지만 오타가 조금씩 있는거 같은데 vscode의 확장기능중에 Code Spell Checker 추천드립니다~ 사용해보시면 생각보다 좋아요! 😄

zwonlala replied:
코드를 짤때 비몽사몽하면서 짰더니 오타가 좀 있네요...!

저도 못 찾은 부분인데 찾아주셔서 감사합니다! :)
좋은 플러그인이네요! 추천 감사합니다 🙌

bingwer commented on src/model/MenuType.js:
```diff
@@ -0,0 +1,44 @@
+	[MenuType.DESERT]: "🍰 디저트",
+});
+
+const MenuTypeUtil = Object.freeze({
```
MenyType의 유틸 함수들을 Object로 관리하셨네요! 유틸함수들이 많아지면 관리하기가 힘들어질때가 있는데 이 방법을 사용하면 유틸함수의 관심사들을 분류할수 있어서 좋은것 같아요! 😀

zwonlala replied:
넵넵 이전에는 TS의 namespace 써서 
저런 enum과 유틸 메소드들을 한 식별자로 묶어줬었는데 
JS에서는 그런 기능이 없는 것 같아 고민하다 보니 
<br>
\~~~Util 이라는 새로운 object 를 정의하고 
\~~~에 관련된 함수들을 해당 object에서 관리해주는 방법으로 해결해보았습니다! 😄 
<br>
피드백 감사합니다!

bingwer reviewed (commented):
리뷰가 좀 많이 늦었네요~ 😂. 코드를 깔끔하게 잘 작성하시기도 했고, 민오님께서 리뷰를 워낙 잘해주셔서 제가 남길게 별로 없었어요.😅
작성하신 코드를 보면서 느낀점은 비즈니스로직을 잘 분류하셔서 데이터의 구조가 잘 잡혀있는것 같아서 좋았습니다! 많이 배워가요!

zwonlala committed:
refactor: EOL 개행 없는 파일들에 EOL 개행 추가

zwonlala committed:
fix typo

zwonlala committed:
docs: update README.md
